src/mpid/ch4/netmod/ucx/ucx_am.h

```diff
@@ -11,52 +11,85 @@
 
 #include "ucx_impl.h"
 
+#define MPIDI_UCX_SHORT_DATA_SZ  16 * 1024
+
 MPL_STATIC_INLINE_PREFIX void MPIDI_UCX_am_isend_callback(void *request, ucs_status_t status)
 {
     MPIDI_UCX_ucp_request_t *ucp_request = (MPIDI_UCX_ucp_request_t *) request;
-    MPIR_Request *req = ucp_request->req;
-    int handler_id = req->dev.ch4.am.netmod_am.ucx.handler_id;
 
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_UCX_AM_ISEND_CALLBACK);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_UCX_AM_ISEND_CALLBACK);
 
-    MPL_free(req->dev.ch4.am.netmod_am.ucx.pack_buffer);
-    req->dev.ch4.am.netmod_am.ucx.pack_buffer = NULL;
-    MPIDIG_global.origin_cbs[handler_id] (req);
-    ucp_request->req = NULL;
+    if (ucp_request->u.am_send.pack_buffer) {
+        MPL_free(ucp_request->u.am_send.pack_buffer);
+    }
+    if (ucp_request->u.am_send.sreq) {
+        MPIDIG_global.origin_cbs[ucp_request->u.am_send.handler_id] (ucp_request->u.am_send.sreq);
+    }
+    MPIDI_UCX_UCP_REQUEST_RELEASE(ucp_request);
 
     MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_UCX_AM_ISEND_CALLBACK);
 }
 
-MPL_STATIC_INLINE_PREFIX void MPIDI_UCX_am_send_callback(void *request, ucs_status_t status)
+/* internal routine for sending data. `send_buf` will be freed on completion. If there is
+   `payload` (when not NULL), an iov of length 2 will be used to avoid extra copy. */
+MPL_STATIC_INLINE_PREFIX int MPIDI_UCX_am_send_internal(int rank, MPIR_Comm * comm, int handler_id,
+                                                        void *send_buf, MPI_Aint send_size,
+                                                        void *payload, MPI_Aint payload_size,
+                                                        MPIR_Request * sreq)
 {
-    MPIDI_UCX_ucp_request_t *ucp_request = (MPIDI_UCX_ucp_request_t *) request;
+    int mpi_errno = MPI_SUCCESS;
 
-    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_UCX_AM_SEND_CALLBACK);
-    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_UCX_AM_SEND_CALLBACK);
+    ucp_ep_h ep = MPIDI_UCX_COMM_TO_EP(comm, rank);
+    uint64_t ucx_tag = MPIDI_UCX_init_tag(0, MPIR_Process.comm_world->rank, MPIDI_UCX_AM_TAG);
 
-    MPL_free(ucp_request->buf);
-    ucp_request->buf = NULL;
-    ucp_request_release(ucp_request);
+    MPIDI_UCX_ucp_request_t *ucp_request;
+    if (payload == NULL) {
+        ucp_request = (void *) ucp_tag_send_nb(ep, send_buf, send_size,
+                                               ucp_dt_make_contig(1), ucx_tag,
+                                               &MPIDI_UCX_am_isend_callback);
+    } else {
+        MPIR_Assert(sreq);
+        ucp_dt_iov_t *iov = MPIDI_UCX_AM_REQUEST(sreq, iov);
+        iov[0].buffer = send_buf;
```

Comment:
This is more than refactoring, no?

Reply:
Originally @gcongiu only did this (iov) in `MPIDI_NM_am_send_reply` because he is only looking at MPI_Send a large message and the large packet is always send in a reply segment. I extended here to all large packets, because the reasoning of using iov applies to just payload size, and it should not care or rely upon upper-layer protocols anyway.

Yeah, there is a bit more than refactor here. Should I change the PR title?

Reply:
Do you have data to justify this change?

Reply:
I really don't like these functional AM changes mixed in with general refactoring. I would prefer the are split out and evaluated in their own PR. I have been bitten by approving refactoring patches lately that actually changed library behavior (and introduced bugs). I'm going to be more strict from now on.

Reply:
hmm... I'm not sure how to split this commit. It's like redraw a decision tree, you can't get the exact behavior with different decision tree.

Let me try remove the am code from this pr altogether